Cloud migration tutorial

jseldess opened the pull request:
A tutorial showing users how to use a local cluster to simulate cross-cloud deployment and then migration from one cloud to another. 

I'm planning to add a simple screencast (with audio or text) demoing this across real cloud platforms, and offering that video as something to watch before you begin. But I'm opening this to get initial feedback on the flow and basic messaging.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/1610)
<!-- Reviewable:end -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/a457025035dfaa8eb2056186993f715bf99a99b6/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/33577695c91c0934e1125ec765253f4da0d31077/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/9de181bb50387f6b60fde752e27f51cfb5f5de5a/

spencerkimball commented on demo-cloud-migration.md:
```diff
@@ -0,0 +1,199 @@
+~~~ shell
+# In a new terminal, start node 1 on cloud 1:
+$ cockroach start --insecure \
+--locality=cloud=platform1 \
```
Why "cloud=platform1" instead of just "cloud=1"?

jseldess replied:
Good point. Done.

spencerkimball commented on demo-cloud-migration.md:
```diff
@@ -0,0 +1,199 @@
+
+## Step 2. Set up HAProxy load balancing
+
+Each CockroachDB node is an equally suitable SQL gateway to your cluster, but to ensure client performance and reliability, it's important to use TCP load balancing. [HAProxy](http://www.haproxy.org/) is one of the most popular open-source TCP load balancers, and CockroachDB includes a built-in command for generating a configuration file that is preset to work with your running cluster, so you'll ...
```
How about: "..., but to ensure an even balancing of client requests across available nodes, we can use a TCP load balancer."

jseldess replied:
Done.

spencerkimball commented on demo-cloud-migration.md:
```diff
@@ -0,0 +1,199 @@
+
+## Step 7. Migrate all data to "cloud 2"
+
+In a new terminal, edit the default replication zone, adding a [hard constraint](configure-replication-zones.html#replication-constraints) that all replicas must be on nodes with `--locality=cloud=platform2`:
```
Why don't you provide some colorful additional narration to these steps that are demonstrating the value proposition. For example: "OK, we're now running three nodes one cloud 1. But what if we'd like to start experimenting with resources provided by another cloud vendor? Let's try that by adding three more nodes to a new cloud platform."

"We now have the cluster replicating across our two simulated cloud vendors. But after experimentation, we're happy with cloud vendor 2, and we decide that we'd like to move everything there. Can we do that without interruption to our live client traffic? Yes, and it's as simple as adding a constraint to our replication configuration that all replicas must be chosen from cloud vendor 2."

You get the idea...we need to give people a scenario so it's not just an abstraction that only folks who've been struggling specifically with these problems will understand.

jseldess replied:
Great feedback. I'll work on this.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/f8645eb6b95b234363c8ca97c85ab28b6092f424/

bdarnell commented:
LGTM

---

Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[demo-cloud-migration.md, line 7 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/1610#-KnBkuBv8ybzkKyoJTfi:-KnBkuBv8ybzkKyoJTfj:bd9b2ui) ([raw file](https://github.com/cockroachdb/docs/blob/f8645eb6b95b234363c8ca97c85ab28b6092f424/demo-cloud-migration.md#L7)):*
> ```Markdown
> ---
> 
> CockroachDB's flexible [replication controls](configure-replication-zones.html) make it trivially easy to run a single CockroachDB cluster across cloud platforms or to migrate a cluster from one cloud to another without any service interruption. This page walks you through a local simulation of the process.
> ```

Cross-cloud migration is just a special case of cross-DC migration. Since this is just a simulation, it seems weird to say it's a simulation of the special case instead of the general case. But I guess that's what gets the clicks...

---

*[demo-cloud-migration.md, line 55 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/1610#-KnBlMR8K6QUuRSa4_9T:-KnBlMR8K6QUuRSa4_9U:b-in9hpr) ([raw file](https://github.com/cockroachdb/docs/blob/f8645eb6b95b234363c8ca97c85ab28b6092f424/demo-cloud-migration.md#L55)):*
> ```Markdown
> 
> ~~~ shell
> $ brew install haproxy
> ```

This is mac-specific. I'm OK with not having instructions for other platforms at this point, but we should say something like "In a new terminal, install HAProxy. If you're on a Mac and using Homebrew, use `brew install haproxy`"

---

*[demo-cloud-migration.md, line 64 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/1610#-KnBlotj0phmUGnk9I1P:-KnBlotj0phmUGnk9I1Q:b8x2bar) ([raw file](https://github.com/cockroachdb/docs/blob/f8645eb6b95b234363c8ca97c85ab28b6092f424/demo-cloud-migration.md#L64)):*
> ```Markdown
> ~~~
> 
> This command generates an `haproxy.cfg` file automatically configured to work with the 3 nodes of your running cluster. In the file, change `bind :26257` to `bind :26270`. This changes the port on which HAProxy accepts requests to a port that is not already in use by a node and that won't be used by the nodes you'll add later.
> ```

These port numbers all start to run together. How about a round number like 26000 instead of 26270?

---

*[demo-cloud-migration.md, line 66 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/1610#-KnBmm9_TEOW4yp1a_YK:-KnBmm9_TEOW4yp1a_YL:b-g7fz9g) ([raw file](https://github.com/cockroachdb/docs/blob/f8645eb6b95b234363c8ca97c85ab28b6092f424/demo-cloud-migration.md#L66)):*
> ```Markdown
> This command generates an `haproxy.cfg` file automatically configured to work with the 3 nodes of your running cluster. In the file, change `bind :26257` to `bind :26270`. This changes the port on which HAProxy accepts requests to a port that is not already in use by a node and that won't be used by the nodes you'll add later.
> 
> ~~~ shell
> ```

`shell` mode doesn't highlight this correctly (it colors `bind` but not any other keywords). Unless there's a more precise option, better to just remove the `shell` tag completely.

---

*[demo-cloud-migration.md, line 98 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/1610#-KnBn9TBThWFEPX3mXhi:-KnBn9TBThWFEPX3mXhj:b-2vuaqr) ([raw file](https://github.com/cockroachdb/docs/blob/f8645eb6b95b234363c8ca97c85ab28b6092f424/demo-cloud-migration.md#L98)):*
> ```Markdown
> 
> ~~~ shell
> $ go get github.com/cockroachdb/loadgen/ycsb
> ```

There should be an introductory section that tells people they'll need to install Go (and we could put the haproxy installation instructions there too)

---

*[demo-cloud-migration.md, line 183 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/1610#-KnBoaVwSS_d1s3pGOSb:-KnBoaVwSS_d1s3pGOSc:b-cqrap5) ([raw file](https://github.com/cockroachdb/docs/blob/f8645eb6b95b234363c8ca97c85ab28b6092f424/demo-cloud-migration.md#L183)):*
> ```Markdown
> ## Step 9. Stop the cluster
> 
> Stop HAProxy and YCSB by switching into their terminals and pressing **CTRL + C**. Do the same for each CockroachDB node.
> ```

Stop YCSB first, or it will start spewing errors as soon as you stop haproxy. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1610#-:-KnBp6kzhhd-2WoHwWoW:bb6ywmx)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[demo-cloud-migration.md, line 64 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/1610#-KnBlotj0phmUGnk9I1P:-KnGKhPlcDSnreaikA8p:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/f8645eb6b95b234363c8ca97c85ab28b6092f424/demo-cloud-migration.md#L64)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

These port numbers all start to run together. How about a round number like 26000 instead of 26270?
</blockquote></details>

Done.

---

*[demo-cloud-migration.md, line 66 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/1610#-KnBmm9_TEOW4yp1a_YK:-KnGKgr2saQYz66BTkFy:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/f8645eb6b95b234363c8ca97c85ab28b6092f424/demo-cloud-migration.md#L66)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`shell` mode doesn't highlight this correctly (it colors `bind` but not any other keywords). Unless there's a more precise option, better to just remove the `shell` tag completely.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1610)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/be3014696c690cb648503a8092024eb017585ca4/

nstewart commented:
Minor feedback (we discussed in person too)
1. it would be useful to have people start in some sort of ./cloud-migration-test directory so they don't get a bunch of data files in their home
2. as written, the ycsb command doesn't work after the go-install. need to include the command to build
3. We shouldn't pre-split the ycsb ranges. I think this breaks (or at least doesn't reinforce) a beginners mental model for how CRDB works. Taking off the split shows new ranges emerging as new load comes in, vs just having it be flat the whole time.
4. In the final 'aha' graph where the load switches we ask users to hover over the graph to see the replica counts, but in my instance all but 1 were zero, which clearly wasn't correct. I think I would just say look at the graph, it's clear from the shape and color change that the load has moved to the other cloud.  For the record, I just realized my crdb local install is beta, but according to Jesse this is still a bug.

jseldess commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[demo-cloud-migration.md, line 55 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/1610#-KnBlMR8K6QUuRSa4_9T:-KnGeISlNiRsNA9VxBaA:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/f8645eb6b95b234363c8ca97c85ab28b6092f424/demo-cloud-migration.md#L55)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This is mac-specific. I'm OK with not having instructions for other platforms at this point, but we should say something like "In a new terminal, install HAProxy. If you're on a Mac and using Homebrew, use `brew install haproxy`"
</blockquote></details>

Done.

---

*[demo-cloud-migration.md, line 98 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/1610#-KnBn9TBThWFEPX3mXhi:-KnGeJqg2AksbbSfHvSg:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/f8645eb6b95b234363c8ca97c85ab28b6092f424/demo-cloud-migration.md#L98)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

There should be an introductory section that tells people they'll need to install Go (and we could put the haproxy installation instructions there too)
</blockquote></details>

Done.

---

*[demo-cloud-migration.md, line 183 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/1610#-KnBoaVwSS_d1s3pGOSb:-KnGLEFR45aTZpy8AT8p:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/f8645eb6b95b234363c8ca97c85ab28b6092f424/demo-cloud-migration.md#L183)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Stop YCSB first, or it will start spewing errors as soon as you stop haproxy. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1610)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ff00ea8aaf016751d72212494e6dd71e2b5a99be/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/2d750409e4ae2c1db9f0532ae914d02774f8b2ea/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/80c4bbc227863b57de4a1ab51cfb40ae72e44007/

jseldess commented:
@nstewart:

- Added a "before you begin" suggestion to create a new directory for running nodes.
- Moved all installations to "before you begin". And since `go get` does in fact build ycsb in `$HOME/go/bin` by default, I've added that path to the command to start ycsb.
- No longer pre-splitting ranges with ycsb. 
- Called out that the replica count on hover will be off for a few minutes but that the graph lines are accurate. Also pointed to the node list view for another accurate view of per-node replica counts.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/02bec3497a9df3647a2347c46e32ae0f850c4359/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/9a770304ed498e4eda92ada7511890ad79edfb18/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/9a770304ed498e4eda92ada7511890ad79edfb18/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/838cb2fb203b365a4b0216d2764931d238704cc4/

jseldess commented:
Merging. Will open a follow-up PR for the screencast part.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/5ae031f0b04d0c645a4c1be7d3752d1723ad5d7e/
